Add "show active tab only" option

canova opened the pull request:
[Example profile](https://deploy-preview-2335--perf-html.netlify.com/public/b919767dac79248386b75350e0000ff8c37ddd2b/calltree/?globalTrackOrder=0-1&localTrackOrderByPid=7738-1-2-0~7790-0-1~&thread=2&v=4)

gregtatum commented on src/components/timeline/index.js:
```diff
@@ -227,22 +286,31 @@ class Timeline extends React.PureComponent<Props> {
-  }),
+  mapStateToProps: state => {
+    const configuration = getProfilerConfiguration(state);
+    const activeBrowsingContextID = configuration
```
This ternary logic is probably worth using a selector for.

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -162,6 +163,27 @@ export function getSearchFilteredMarkerIndexes(
   return newMarkers;
 }
 
+export function getTabFilteredMarkerIndexes(
```
We should probably audit some special cases here. I'd probably want to retain "global" things, like the responsiveness markers.

gregtatum replied:
I think maybe we should file a follow-up for this point. Is there already one on file?

gregtatum replied:
Re-resolving.

gregtatum commented on src/selectors/per-thread/thread.js:
```diff
@@ -116,8 +116,14 @@ export function getThreadSelectorsPerThread(threadIndex: ThreadIndex): * {
   );
 
-  const getFilteredThread: Selector<Thread> = createSelector(
+  const _getTabFilteredThread: Selector<Thread> = createSelector(
```
This only drops samples right? It doesn't modify stacks. I haven't really looked too closely at the implementation, but I think it's important to consider a couple of things:

* We can drop samples, and it doesn't really change any of our filter pipelines
* Change stack shapes is very dangerous, as it affects our CallNodePaths. If we plan to do this, we should think through where this happens in our data transform pipeline.
* If we put the tab filtering too late in the pipeline, it affects the thread activity graph.

For instance filtering by active tab produces:

<img width="857" alt="image" src="https://user-images.githubusercontent.com/1588648/70663928-fc964000-1c2e-11ea-8f38-d75e2791e781.png">

Which is great for the gecko engineer workflow, but not great for the webdev workflow. It would be better to show the graphs as if they were not doing work there.

gregtatum commented on src/components/timeline/index.js:
```diff
@@ -153,6 +161,49 @@ class TimelineSettingsHiddenTracks extends React.PureComponent<{|
+    }
+
+    return (
+      <form>
```
We should probably only show this checkbox for non-webdev presets.

julienw commented on src/app-logic/url-handling.js:
```diff
@@ -365,6 +368,7 @@ export function stateFromLocation(
     selectedTab: toValidTabSlug(pathParts[selectedTabPathPart]) || 'calltree',
     pathInZipFile: query.file || null,
     profileName: query.profileName,
+    showTabOnly: query.showTabOnly ? Number(query.showTabOnly) : null,
```
maybe check that the result of the cast to a number is correct (`Number.isInteger` could help here)

julienw commented on src/components/timeline/index.js:
```diff
@@ -227,22 +286,31 @@ class Timeline extends React.PureComponent<Props> {
+  mapStateToProps: state => {
+    const configuration = getProfilerConfiguration(state);
+    const activeBrowsingContextID = configuration
+      ? configuration.activeBrowsingContextID || null
```
This is a dangerous use of `||` with an integer: is it possible that its value is `0` ?

canova replied:
`0` means we failed to get the BrowsingContext ID from the browser. The job of this `||`  is precisely convert `0` to `null`. Adding a comment that explains this.

canova replied:
Yes, I added a comment to this code(and moved this to a selector). Please see `src/selectors/profile.js` file. And sorry about the outdated codes after some commits. I wanted to rebase everything to appropriate commits but there were a lot of changes, and it was really hard to make :/

julienw commented on src/selectors/profile.js:
```diff
@@ -447,3 +451,43 @@ export const getHiddenTrackCount: Selector<HiddenTrackCount> = createSelector(
+        }
+      }
+      return page;
+    };
```
The algorithm doesn't seem good enough to me. It's basically n^2 because for each page in pageList we look in pageList again. It's OK if we know that pageList is always small, but can we assume that?
Can we order pageList in such a way that we know that parents immediately come before or after (like we do for our other structures) so that we don't need to look at it completely? Otherwise it might be a good idea to use a Map with innerWindowID as key so that we loop the list only once, and then the lookup is constant time.
The strategy depends greatly on the size of the pageList structure obviously.

Also this selector could use some comments, explaining better the structure of the list and its content (that it's actually a tree, etc).

gregtatum replied:
@canova does this comment still need handling?

digitarald commented on src/components/timeline/index.js:
```diff
@@ -153,6 +161,49 @@ class TimelineSettingsHiddenTracks extends React.PureComponent<{|
+    return (
+      <form>
+        <div className="timelineSettingsToggle">
+          <label className="photon-label photon-label-micro timelineSettingsToggleLabel">
```
Could we size the checkbox to match the checkbox in the _Invert call stack_ in the _Call Tree_ toolbar. It seems larger.

gregtatum commented on src/profile-logic/profile-data.js:
```diff
@@ -1027,6 +1027,53 @@ export function filterThreadToSearchString(
   );
 }
 
+export function filterThreadByTab(
```
Please add a documentation block to the top of this function. Also, I had a bit of trouble reading through the implementation. I think it would be a good idea to try and explain the intent of what the function is doing with some inline comments as well.

gregtatum commented on src/profile-logic/profile-data.js:
```diff
@@ -1027,6 +1027,53 @@ export function filterThreadToSearchString(
+
+    const { frameTable, stackTable } = thread;
+
+    const frameMatchesFilterCache = new Map();
```
It would probably be helpful to add an inline type here, I find the `Map` type hard to reason with when it's implicit like this, and I think it's easy to use wrong and have things slip in there, like `null` and `undefined`.

gregtatum commented on src/profile-logic/profile-data.js:
```diff
@@ -1027,6 +1027,53 @@ export function filterThreadToSearchString(
+      let result = frameMatchesFilterCache.get(frame);
+      if (result === undefined) {
+        const innerWindowID = frameTable.innerWindowID[frame];
+        result = innerWindowID && relevantPages.has(innerWindowID);
```
I find the `&&` operators extremely hard to reason with. Would you mind separating out the condition and the assignment? I usually find the ternary is much easier to read.

gregtatum replied:
Also, please add some strict equality checking here. The implicit is dangerous, especially as there are no comments here explaining things. I THINK innerWindowID will be 0 when it can't find one.

gregtatum commented on src/profile-logic/profile-data.js:
```diff
@@ -1027,6 +1027,53 @@ export function filterThreadToSearchString(
+
+    const frameMatchesFilterCache = new Map();
+    function frameMatchesFilter(frame) {
+      let result = frameMatchesFilterCache.get(frame);
```
It's also probably a good idea to add an implicit type to a `let` variable. It takes a lot of reasoning to figure this one out here, especially since `Map` is also not typed.

gregtatum commented on src/selectors/profile.js:
```diff
@@ -450,3 +451,43 @@ export const getHiddenTrackCount: Selector<HiddenTrackCount> = createSelector(
   }
 );
+
+export const getRelevantPagesForActiveTab: Selector<
```
Please add some documentation for this function. I think it would be good to include documentation of what makes a page relevant.

gregtatum replied:
I don't see a specific test for this function. It's implementing some tricky logic, so I think a test for it would be a good idea.

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -162,6 +163,27 @@ export function getSearchFilteredMarkerIndexes(
   return newMarkers;
 }
 
+export function getTabFilteredMarkerIndexes(
```
Please add a code comment explaining this selector, it would be nice for someone who's not familiar with the feature.

gregtatum commented on src/selectors/per-thread/markers.js:
```diff
@@ -186,16 +186,28 @@ export function getMarkerSelectorsPerThread(threadSelectors: *) {
   );
 
+  /**
+   * This selector applies the committed range to the full list of markers.
```
I think this code comment could be a bit more accurate, especially around "full list of markers".

gregtatum commented on src/selectors/profile.js:
```diff
@@ -154,6 +155,17 @@ export const getContentfulSpeedIndexProgress: Selector<
+    // BrowsingContext ID can be `0` and that means Firefox has failed to get
+    // the BrowsingContextID of the active tab. We are converting that `0` to
+    // `null` here to explicitly indicate that we don't have that information.
+    return configuration.activeBrowsingContextID || null;
```
Thanks for the explanation here. I'm still a bit nervous using truthy values to compute this. I'd be more comfortable with explicit checks.

gregtatum commented on src/selectors/profile.js:
```diff
@@ -503,3 +503,40 @@ export const getRelevantPagesForActiveTab: Selector<
   }
 );
+
+export const getIsLocalTrackHidden: DangerousSelectorWithArguments<
```
@julienw I'm happy we came to this type name `DangerousSelectorWithArguments`.

gregtatum commented on src/selectors/profile.js:
```diff
@@ -503,3 +503,40 @@ export const getRelevantPagesForActiveTab: Selector<
+      case 'network':
+      case 'memory':
+      case 'ipc':
+        // Hide those local track types because we want to hide as much as
```
Thanks for the explanation here.

gregtatum commented on src/profile-logic/profile-data.js:
```diff
@@ -1027,6 +1034,59 @@ export function filterThreadToSearchString(
+  return timeCode('filterThreadByTab', () => {
+    if (relevantPages.size === 0) {
+      // Either there is no relevant page or "active tab only" view is not active.
+      return thread;
```
Optional: It might be worth testing that this properly memoizes, and returns the original thread if no work is needed to be done here. Especially as `getSamplesSelectedStates` takes the result of this as the argument.

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -174,7 +174,16 @@ export function getTabFilteredMarkerIndexes(
-    const { data } = getMarker(markerIndex);
+    const { name, data } = getMarker(markerIndex);
+
+    // We want to retain some markers even though they do not belong to a specific tab.
```
Ah, here is this code. I commented earlier about this.

gregtatum commented on src/profile-logic/marker-data.js:
```diff
@@ -166,7 +166,8 @@ export function getSearchFilteredMarkerIndexes(
   markerIndexes: MarkerIndex[],
-  relevantPages: Set<InnerWindowID>
+  relevantPages: Set<InnerWindowID>,
+  includeGlobalMarkers: boolean = true
```
Does this need to be configurable?

gregtatum commented on src/selectors/app.js:
```diff
@@ -52,6 +56,210 @@ export const getTrackThreadHeights: Selector<
 export const getIsNewlyPublished: Selector<boolean> = state =>
   getApp(state).isNewlyPublished;
 
+export const getComputedHiddenGlobalTracks: Selector<
```
TODO - I'd like to test this more. Please don't resolve this comment until I get a chance to. I don't know if I'll have time today.

gregtatum replied:
Please add a documentation comment at the top of the function.

gregtatum commented on src/selectors/app.js:
```diff
@@ -52,6 +56,210 @@ export const getTrackThreadHeights: Selector<
+  ) => {
+    if (!showTabOnly) {
+      // Return the usual hidden thread if we are not in the "single tab view" mode.
+      return hiddenGlobalTracks;
```
Ah, we'll only incur the cost if the showTabOnly is checked, which is good.

gregtatum commented on src/selectors/app.js:
```diff
@@ -52,6 +56,210 @@ export const getTrackThreadHeights: Selector<
+      const globalTrack = globalTracks[trackIndex];
+      const trackType = globalTrack.type;
+      switch (trackType) {
+        case 'screenshots':
```
Can we keep screenshots for now, as they are still really useful?

gregtatum commented on src/selectors/app.js:
```diff
@@ -52,6 +56,210 @@ export const getTrackThreadHeights: Selector<
+  }
+);
+
+export const getComputedHiddenLocalTracksByPid: Selector<
```
Please add a code comment at the top of this selector.

gregtatum commented on src/test/store/tracks.test.js:
```diff
@@ -603,6 +607,25 @@ describe('ordering and hiding', function() {
+        '  - show [thread Style]',
+      ]);
+      // Select an existing BrowsingContextID
+      dispatch(changeShowTabOnly(123123));
```
Is there a better way to get this magic number here?

gregtatum commented on src/test/store/tracks.test.js:
```diff
@@ -603,6 +607,25 @@ describe('ordering and hiding', function() {
+      const { getState, dispatch, tabThreadIndex } = init();
+      dispatch(changeSelectedThread(tabThreadIndex));
+      expect(getHumanReadableTracks(getState())).toEqual([
+        'show [thread GeckoMain process]',
```
Thanks for the tests here.

gregtatum commented on src/types/actions.js:
```diff
@@ -361,6 +361,7 @@ type UrlStateAction =
   | {|
       +type: 'CHANGE_SHOW_TAB_ONLY',
       +showTabOnly: BrowsingContextID | null,
+      +selectedTab: TabSlug,
```
Not for this patch: I think I did similar approaches elsewhere when tabs disappear, but I think this logic for switching to a supported tab could be placed in a selector.

gregtatum commented on src/test/store/profile-view.test.js:
```diff
@@ -1344,6 +1370,14 @@ describe('snapshots of selectors/profile', function() {
       'Merge: C',
     ]);
   });
+  it('matches the last stored run of selectedThreadSelector.getTabFilteredThread', function() {
```
Optional: This will get the job done quick, but it might be better to write non-snapshot tests.

canova commented on src/components/timeline/index.js:
```diff
@@ -153,6 +161,49 @@ class TimelineSettingsHiddenTracks extends React.PureComponent<{|
+    }
+
+    return (
+      <form>
```
Well, it doesn't do anything :) Wanted to keep a form element there because other elements also had their `form` elements, and thought it would be good to keep the same convention. But I can remove it

gregtatum commented:
So I think we should mitigate this issue:

[Deploy preview](https://deploy-preview-2335--perf-html.netlify.com/public/90e838a880c9b2c56c4ef9a7464c237c618a05e1/calltree/?globalTrackOrder=0-1-2-3-4-5&hiddenGlobalTracks=1-2-3-4-5&localTrackOrderByPid=84476-1-2-0~84501-0~84498-0~84523-0~84500-0~84764-0-1~&showTabOnly=16&thread=0&v=4)
<img width="1552" alt="image" src="https://user-images.githubusercontent.com/1588648/72925141-ce208000-3d17-11ea-8d26-d89281caeec9.png">

gregtatum commented:
Here's a similar issue:

https://perfht.ml/2NNIxL6

<img width="285" alt="image" src="https://user-images.githubusercontent.com/1588648/72925722-cd3c1e00-3d18-11ea-800e-644115cb636b.png">

STR:
 * View the profile that is showing the active tab.
 * Hide Web Content 5/5

Expected results:
 * It won't hide the last track.

Actual results
 * The track gets hidden.

gregtatum commented:
The "show active tab only" mechanism is bypassing the hidden track settings.

<img width="352" alt="image" src="https://user-images.githubusercontent.com/1588648/72925924-2015d580-3d19-11ea-9545-756559a458a1.png">

So the UI is not consistent with what's being displayed to the user.

gregtatum commented:
Feel free to hit me up synchronously on this, but I think there is an issue with the approach on hiding tracks, outlined above. What do you think about landing the checkbox first, but then follow-up with the track management? I think it would still be useful without the track hiding. It would also simplify this review process.

I wish I had thought of this when you were first working on this problem, but I think the following approach will create a more consistent experience for our users. It pre-computes the list of active tab tracks, then uses that information in the selectors to handle the variety of cases correctly. The computation would go in the finalizeProfile section along with the rest of the track computations.

```diff
diff --git a/src/actions/receive-profile.js b/src/actions/receive-profile.js
index 6689bf38..489e2cb2 100644
--- a/src/actions/receive-profile.js
+++ b/src/actions/receive-profile.js
@@ -164,6 +164,12 @@ export function finalizeProfileView(
       hasUrlInfo ? getHiddenGlobalTracks(getState()) : null,
       getLegacyHiddenThreads(getState())
     );
+    // Pre-compute which tracks are available for the active tab.
+    let activeTabGlobalTracks = computeActiveTabGlobalTracks(
+      ...whateverYouNeed,
+      // The state can be passed to this as well, so you can run any selector you need.
+      getState()
+    );
     const localTracksByPid = computeLocalTracksByPid(profile);
     const localTrackOrderByPid = initializeLocalTrackOrderByPid(
       hasUrlInfo ? getLocalTrackOrderByPid(getState()) : null,
@@ -176,6 +182,13 @@ export function finalizeProfileView(
       profile,
       getLegacyHiddenThreads(getState())
     );
+    // Pre-compute which local tracks are available for the active tab. Then in the
+    // selectors.
+    let activeTabLocalTracksByPid = computeActiveTabLocalTracksByPid(
+      localTracksByPid,
+      // The state can be passed to this as well, so you can run any selector you need.
+      getState()
+    );
     let visibleThreadIndexes = getVisibleThreads(
       globalTracks,
       hiddenGlobalTracks,
@@ -235,6 +248,10 @@ export function finalizeProfileView(
       localTracksByPid,
       hiddenLocalTracksByPid,
       localTrackOrderByPid,
+      // Save these in a selectors/profile-view.js, and then pass them to the selectors
+      // so that they can do the right thing.
+      activeTabGlobalTracks,
+      activeTabLocalTracksByPid,
     });
 
     // Note we kick off symbolication only for the profiles we know for sure
diff --git a/src/components/timeline/TrackContextMenu.js b/src/components/timeline/TrackContextMenu.js
index 4baa74fe..4e4dab45 100644
--- a/src/components/timeline/TrackContextMenu.js
+++ b/src/components/timeline/TrackContextMenu.js
@@ -196,10 +196,22 @@ class TimelineTrackContextMenu extends PureComponent<Props> {
   };
 
   renderGlobalTrack(trackIndex: TrackIndex) {
-    const { hiddenGlobalTracks, globalTrackNames, globalTracks } = this.props;
+    const {
+      showTabOnly,
+      activeTabGlobalTracks,
+      hiddenGlobalTracks,
+      globalTrackNames,
+      globalTracks,
+    } = this.props;
     const isHidden = hiddenGlobalTracks.has(trackIndex);
     const track = globalTracks[trackIndex];
 
+    if (showTabOnly && !activeTabGlobalTracks.has(trackIndex)) {
+      // Handle the checkbox that doesn't have a global track. Also do this for local
+      // tracks.
+      return null;
+    }
+
     let title = `${globalTrackNames[trackIndex]}`;
     if (track.type === 'process') {
       title += ` (Process ID: ${track.pid})`;
diff --git a/src/selectors/app.js b/src/selectors/app.js
index 53b8c56c..a81686e3 100644
--- a/src/selectors/app.js
+++ b/src/selectors/app.js
@@ -56,6 +56,7 @@ export const getTrackThreadHeights: Selector<
 export const getIsNewlyPublished: Selector<boolean> = state =>
   getApp(state).isNewlyPublished;
 
+// This probably goes away:
 export const getComputedHiddenGlobalTracks: Selector<
   Set<TrackIndex>
 > = createSelector(
diff --git a/src/selectors/url-state.js b/src/selectors/url-state.js
index b46b78c9..bd5c48e8 100644
--- a/src/selectors/url-state.js
+++ b/src/selectors/url-state.js
@@ -88,8 +88,19 @@ export const getLegacyHiddenThreads: Selector<ThreadIndex[] | null> = state =>
   getProfileSpecificState(state).legacyHiddenThreads;
 export const getGlobalTrackOrder: Selector<TrackIndex[]> = state =>
   getProfileSpecificState(state).globalTrackOrder;
-export const getHiddenGlobalTracks: Selector<Set<TrackIndex>> = state =>
-  getProfileSpecificState(state).hiddenGlobalTracks;
+export const getHiddenGlobalTracks: Selector<Set<TrackIndex>> = createSelector(
+  getProfileSpecificState,
+  getShowTabOnly,
+  getActiveTabGlobalTracks,
+  (profileSpecific, showTabOnly, activeTabGlobalTracks) => {
+    const hiddenGlobalTracks = getProfileSpecificState(state)
+      .hiddenGlobalTracks;
+    if (!showTabOnly) {
+      return hiddenGlobalTracks;
+    }
+    // Handle this.
+  }
+);
 export const getHiddenLocalTracksByPid: Selector<
   Map<Pid, Set<TrackIndex>>
 > = state => getProfileSpecificState(state).hiddenLocalTracksByPid;
```

canova commented:
@gregtatum Addressed your review comments and removed the thread hiding logic from this PR, will open a follow up PR with your suggestion. Could you take a look at the PR again when you have time?

gregtatum commented:
I didn't really have the mental capacity to handle this review today, but I did check out the UI changes a bit. I think it would be good to handle the context menu here a bit more gracefully. Happy to chat on Monday more and take a look at the code changes more in depth.

![ui-example](https://user-images.githubusercontent.com/1588648/74066924-b1de3d80-49bd-11ea-901b-004b4ae1a0db.gif)

canova commented:
@gregtatum Changed that context menu and added some small tests. It should be working fine now. Could you take a look at it again when you have some time?

canova commented:
Thanks for the review Greg! We decided to keep the Jank markers even thought they don't have an innerWindowID in them since we believe they are important for web developers as well. But we should hide the empty tracks(even though they have some Jank markers in it, they are technically empty). I will work on hiding the empty tracks next, which should solve this issue as well. Filed an issue for that: #2403.
